kafka exporter: panic: reflect.Set: value of type string is not assignable to type config.Secret

### What's wrong?

panic

```
panic: reflect.Set: value of type string is not assignable to type config.Secret
```

no `sasl_password` covered in tests https://github.com/grafana/agent/blob/main/component/prometheus/exporter/kafka/kafka_test.go

### Steps to reproduce


example configuration

```
prometheus.exporter.kafka "kafka_hetzner" {
    kafka_uris = ["broker1"]
    use_sasl = true
    sasl_password = "foobar" // <-- here
}
```

### System information

_No response_

### Software version

0.38.1

### Configuration

```
prometheus.exporter.kafka "kafka_hetzner" {
    kafka_uris = ["broker1"]
    use_sasl = true
    sasl_password = "foobar" // <-- here
}
```

### Logs

```
ts=2024-01-04T11:08:34.175127661Z level=info msg="finished node evaluation" controller_id="" trace_id=4759262a337e25869bad3076f7596e8b node_id=http duration=23.701µs
ts=2024-01-04T11:08:34.175153422Z level=info msg="finished node evaluation" controller_id="" trace_id=4759262a337e25869bad3076f7596e8b node_id=ui duration=869ns
ts=2024-01-04T11:08:34.175243388Z level=info msg="finished node evaluation" controller_id="" trace_id=4759262a337e25869bad3076f7596e8b node_id=prometheus.exporter.kafka.kafka_hetzner duration=68.756µs
ts=2024-01-04T11:08:34.175280922Z level=info msg="finished complete graph evaluation" controller_id="" trace_id=4759262a337e25869bad3076f7596e8b duration=438.628µs
panic: reflect.Set: value of type string is not assignable to type config.Secret

goroutine 1 [running]:
reflect.Value.assignTo({0x60e8ac0?, 0xc002953220?, 0x7fdf4c339b38?}, {0x773f8e5, 0xb}, 0x66846a0, 0x0)
        /usr/local/go/src/reflect/value.go:3307 +0x289
reflect.Value.Set({0x66846a0?, 0xc00290aa40?, 0x6791a00?}, {0x60e8ac0?, 0xc002953220?, 0x66846a0?})
        /usr/local/go/src/reflect/value.go:2260 +0xe6
github.com/grafana/river/internal/value.(*decoder).decode(0xc002b861b7, {{0x60e8ac0?, 0xc002953210?, 0x412a09?}, 0x88?}, {0x66846a0, 0xc00290aa40, 0x198})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/internal/value/decode.go:215 +0x1266
github.com/grafana/river/internal/value.Decode({{0x60e8ac0?, 0xc002953210?, 0xc00297f0e0?}, 0xa8?}, {0x68cf9c0?, 0xc00290aa40?})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/internal/value/decode.go:80 +0xed
github.com/grafana/river/vm.(*structDecoder).decodeAttr(0xc00257e088, 0xc0029741c8, {0x758e100?, 0xc00290aa00?, 0x13?}, 0xc00257d9f0)
        /go/pkg/mod/github.com/grafana/river@v0.3.0/vm/struct_decoder.go:181 +0x75d
github.com/grafana/river/vm.(*structDecoder).Decode(0xc00257e088, {0xc00283b900, 0x8, 0x68ef4c0?}, {0x758e100?, 0xc00290aa00?, 0x5b31563?})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/vm/struct_decoder.go:71 +0x726
github.com/grafana/river/vm.(*Evaluator).evaluateDecode(0xc002952bf0, 0xc002953020, 0xc00297f0e0, {0x87a7f00?, 0xc002974378?}, {0x68ef4c0?, 0xc00290aa00?, 0x30?})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/vm/vm.go:169 +0x34d
github.com/grafana/river/vm.(*Evaluator).evaluateBlockOrBody(0x128?, 0x758e100?, 0xc00257e188?, {0x87a7f00, 0xc002974378}, {0x68ef4c0?, 0xc00290aa00?, 0xc00290aa00?})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/vm/vm.go:93 +0x10d
github.com/grafana/river/vm.(*Evaluator).Evaluate(0xc002952bf0, 0x19?, {0x68ef4c0, 0xc00290aa00})
        /go/pkg/mod/github.com/grafana/river@v0.3.0/vm/vm.go:58 +0x17b
github.com/grafana/agent/pkg/flow/internal/controller.(*ComponentNode).evaluate(0xc002934480, 0x0?)
        /src/agent/pkg/flow/internal/controller/node_component.go:268 +0x130
github.com/grafana/agent/pkg/flow/internal/controller.(*ComponentNode).Evaluate(0xc00276b8b0?, 0xcf3f280?)
        /src/agent/pkg/flow/internal/controller/node_component.go:251 +0x1c
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).evaluate(0xc002391800, {0x87a44c0, 0xc00297c410}, {0x87fe1a0?, 0xc002934480?})
        /src/agent/pkg/flow/internal/controller/loader.go:701 +0x58
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply.func2({0x87a7fa0, 0xc002934480})
        /src/agent/pkg/flow/internal/controller/loader.go:177 +0x585
github.com/grafana/agent/pkg/flow/internal/dag.WalkTopological(0xc00297a1c0, {0xc00283bd00, 0x5, 0xcbf79f0?}, 0xc00257f038)
        /src/agent/pkg/flow/internal/dag/walk.go:83 +0x226
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply(0xc002391800, 0x2a0?, {0xc00297a1a0?, 0x3, 0x4}, {0xc002736658, 0x1, 0x1})
        /src/agent/pkg/flow/internal/controller/loader.go:160 +0xaad
github.com/grafana/agent/pkg/flow.(*Flow).LoadSource(0xc002172870, 0xc002976300, 0x7725e6b?)
        /src/agent/pkg/flow/flow.go:283 +0xd6
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run.func4()
        /src/agent/cmd/internal/flowmode/cmd_run.go:273 +0x268
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run(0xc002874540, {0x7fff097166b5, 0x17})
        /src/agent/cmd/internal/flowmode/cmd_run.go:306 +0xea2
github.com/grafana/agent/cmd/internal/flowmode.runCommand.func1(0xc002837b00?, {0xc0029301c0?, 0x4?, 0x7725e9f?})
        /src/agent/cmd/internal/flowmode/cmd_run.go:93 +0x2b
github.com/spf13/cobra.(*Command).execute(0xc002345800, {0xc002930150, 0x7, 0x7})
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0xc002344f00)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/grafana/agent/cmd/internal/flowmode.Run()
        /src/agent/cmd/internal/flowmode/flowmode.go:33 +0x305
main.main()
        /src/agent/cmd/grafana-agent/main.go:50 +0x6bd
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,8 +14,8 @@ Main (unreleased)
 
 - `otelcol.receiver.prometheus` will drop all `otel_scope_info` metrics when converting them to OTLP. (@wildum)
   - If the `otel_scope_info` metric has labels `otel_scope_name` and `otel_scope_version`,
-    their values will be used to set OTLP Instrumentation Scope name and  version respectively. 
-  - Labels of `otel_scope_info` metrics other than `otel_scope_name` and `otel_scope_version` 
+    their values will be used to set OTLP Instrumentation Scope name and  version respectively.
+  - Labels of `otel_scope_info` metrics other than `otel_scope_name` and `otel_scope_version`
     are added as scope attributes with the matching name and version.
 
 - The `target` block in `prometheus.exporter.blackbox` requires a mandatory `name`
@@ -57,7 +57,7 @@ Main (unreleased)
 
 - Added links between compatible components in the documentation to make it
   easier to discover them. (@thampiotr)
-  
+
 - Allow defining `HTTPClientConfig` for `discovery.ec2`. (@cmbrad)
 
 - The `remote.http` component can optionally define a request body. (@tpaschalis)
@@ -71,8 +71,8 @@ Main (unreleased)
 - Added 'country' mmdb-type to log pipeline-stage geoip. (@superstes)
 
 - Azure exporter enhancements for flow and static mode, (@kgeckhart)
-  - Allows for pulling metrics at the Azure subscription level instead of resource by resource 
-  - Disable dimension validation by default to reduce the number of exporter instances needed for full dimension coverage 
+  - Allows for pulling metrics at the Azure subscription level instead of resource by resource
+  - Disable dimension validation by default to reduce the number of exporter instances needed for full dimension coverage
 
 - Add `max_cache_size` to `prometheus.relabel` to allow configurability instead of hard coded 100,000. (@mattdurham)
 
@@ -81,11 +81,11 @@ Main (unreleased)
 ### Bugfixes
 
 - Update `pyroscope.ebpf` to fix a logical bug causing to profile to many kthreads instead of regular processes https://github.com/grafana/pyroscope/pull/2778 (@korniltsev)
- 
+
 - Update `pyroscope.ebpf` to produce more optimal pprof profiles for python processes https://github.com/grafana/pyroscope/pull/2788 (@korniltsev)
 
 - In Static mode's `traces` subsystem, `spanmetrics` used to be generated prior to load balancing.
-  This could lead to inaccurate metrics. This issue only affects Agents using both `spanmetrics` and 
+  This could lead to inaccurate metrics. This issue only affects Agents using both `spanmetrics` and
   `load_balancing`, when running in a load balanced cluster with more than one Agent instance. (@ptodev)
 
 - Fixes `loki.source.docker` a behavior that synced an incomplete list of targets to the tailer manager. (@FerdinandvHagen)
@@ -93,6 +93,8 @@ Main (unreleased)
 - Fixes `otelcol.connector.servicegraph` store ttl default value from 2ms to 2s. (@rlankfo)
 
 - Add staleness tracking to labelstore to reduce memory usage. (@mattdurham)
+
+- Fix issue where `prometheus.exporter.kafka` would crash when configuring `sasl_password`. (@rfratto)
 
 ### Other changes
 

--- a/component/prometheus/exporter/kafka/kafka.go
+++ b/component/prometheus/exporter/kafka/kafka.go
@@ -9,7 +9,8 @@ import (
 	"github.com/grafana/agent/component/prometheus/exporter"
 	"github.com/grafana/agent/pkg/integrations"
 	"github.com/grafana/agent/pkg/integrations/kafka_exporter"
-	config_util "github.com/prometheus/common/config"
+	"github.com/grafana/river/rivertypes"
+	"github.com/prometheus/common/config"
 )
 
 var DefaultArguments = Arguments{
@@ -24,28 +25,28 @@ var DefaultArguments = Arguments{
 }
 
 type Arguments struct {
-	Instance                string             `river:"instance,attr,optional"`
-	KafkaURIs               []string           `river:"kafka_uris,attr,optional"`
-	UseSASL                 bool               `river:"use_sasl,attr,optional"`
-	UseSASLHandshake        bool               `river:"use_sasl_handshake,attr,optional"`
-	SASLUsername            string             `river:"sasl_username,attr,optional"`
-	SASLPassword            config_util.Secret `river:"sasl_password,attr,optional"`
-	SASLMechanism           string             `river:"sasl_mechanism,attr,optional"`
-	UseTLS                  bool               `river:"use_tls,attr,optional"`
-	CAFile                  string             `river:"ca_file,attr,optional"`
-	CertFile                string             `river:"cert_file,attr,optional"`
-	KeyFile                 string             `river:"key_file,attr,optional"`
-	InsecureSkipVerify      bool               `river:"insecure_skip_verify,attr,optional"`
-	KafkaVersion            string             `river:"kafka_version,attr,optional"`
-	UseZooKeeperLag         bool               `river:"use_zookeeper_lag,attr,optional"`
-	ZookeeperURIs           []string           `river:"zookeeper_uris,attr,optional"`
-	ClusterName             string             `river:"kafka_cluster_name,attr,optional"`
-	MetadataRefreshInterval string             `river:"metadata_refresh_interval,attr,optional"`
-	AllowConcurrent         bool               `river:"allow_concurrency,attr,optional"`
-	MaxOffsets              int                `river:"max_offsets,attr,optional"`
-	PruneIntervalSeconds    int                `river:"prune_interval_seconds,attr,optional"`
-	TopicsFilter            string             `river:"topics_filter_regex,attr,optional"`
-	GroupFilter             string             `river:"groups_filter_regex,attr,optional"`
+	Instance                string            `river:"instance,attr,optional"`
+	KafkaURIs               []string          `river:"kafka_uris,attr,optional"`
+	UseSASL                 bool              `river:"use_sasl,attr,optional"`
+	UseSASLHandshake        bool              `river:"use_sasl_handshake,attr,optional"`
+	SASLUsername            string            `river:"sasl_username,attr,optional"`
+	SASLPassword            rivertypes.Secret `river:"sasl_password,attr,optional"`
+	SASLMechanism           string            `river:"sasl_mechanism,attr,optional"`
+	UseTLS                  bool              `river:"use_tls,attr,optional"`
+	CAFile                  string            `river:"ca_file,attr,optional"`
+	CertFile                string            `river:"cert_file,attr,optional"`
+	KeyFile                 string            `river:"key_file,attr,optional"`
+	InsecureSkipVerify      bool              `river:"insecure_skip_verify,attr,optional"`
+	KafkaVersion            string            `river:"kafka_version,attr,optional"`
+	UseZooKeeperLag         bool              `river:"use_zookeeper_lag,attr,optional"`
+	ZookeeperURIs           []string          `river:"zookeeper_uris,attr,optional"`
+	ClusterName             string            `river:"kafka_cluster_name,attr,optional"`
+	MetadataRefreshInterval string            `river:"metadata_refresh_interval,attr,optional"`
+	AllowConcurrent         bool              `river:"allow_concurrency,attr,optional"`
+	MaxOffsets              int               `river:"max_offsets,attr,optional"`
+	PruneIntervalSeconds    int               `river:"prune_interval_seconds,attr,optional"`
+	TopicsFilter            string            `river:"topics_filter_regex,attr,optional"`
+	GroupFilter             string            `river:"groups_filter_regex,attr,optional"`
 }
 
 func init() {
@@ -93,7 +94,7 @@ func (a *Arguments) Convert() *kafka_exporter.Config {
 		UseSASL:                 a.UseSASL,
 		UseSASLHandshake:        a.UseSASLHandshake,
 		SASLUsername:            a.SASLUsername,
-		SASLPassword:            a.SASLPassword,
+		SASLPassword:            config.Secret(a.SASLPassword),
 		SASLMechanism:           a.SASLMechanism,
 		UseTLS:                  a.UseTLS,
 		CAFile:                  a.CAFile,

--- a/component/prometheus/exporter/kafka/kafka_test.go
+++ b/component/prometheus/exporter/kafka/kafka_test.go
@@ -107,3 +107,15 @@ func TestCustomizeTarget(t *testing.T) {
 	require.Equal(t, 1, len(newTargets))
 	require.Equal(t, "example", newTargets[0]["instance"])
 }
+
+func TestSASLPassword(t *testing.T) { // #6044
+	var exampleRiverConfig = `
+		kafka_uris    = ["broker1"]
+		use_sasl      = true 
+		sasl_password = "foobar"
+	`
+
+	var args Arguments
+	err := river.Unmarshal([]byte(exampleRiverConfig), &args)
+	require.NoError(t, err)
+}

--- a/converter/internal/staticconvert/internal/build/kafka_exporter.go
+++ b/converter/internal/staticconvert/internal/build/kafka_exporter.go
@@ -4,6 +4,7 @@ import (
 	"github.com/grafana/agent/component/discovery"
 	"github.com/grafana/agent/component/prometheus/exporter/kafka"
 	"github.com/grafana/agent/pkg/integrations/kafka_exporter"
+	"github.com/grafana/river/rivertypes"
 )
 
 func (b *IntegrationsConfigBuilder) appendKafkaExporter(config *kafka_exporter.Config, instanceKey *string) discovery.Exports {
@@ -17,7 +18,7 @@ func toKafkaExporter(config *kafka_exporter.Config) *kafka.Arguments {
 		UseSASL:                 config.UseSASL,
 		UseSASLHandshake:        config.UseSASLHandshake,
 		SASLUsername:            config.SASLUsername,
-		SASLPassword:            config.SASLPassword,
+		SASLPassword:            rivertypes.Secret(config.SASLPassword),
 		SASLMechanism:           config.SASLMechanism,
 		UseTLS:                  config.UseTLS,
 		CAFile:                  config.CAFile,

